verilog/analysis/checkers/forbid_implicit_declarations_rule_test.cc

```diff
@@ -0,0 +1,245 @@
+       "  assign a1 = 1'b0;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+       "  module foo;\n"
+       "  endmodule\n"
+       "endmodule"},
+      {"module m;\n"
+       "  module foo;\n"
+       "  endmodule\n"
+       "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  wire a1;\n"
+       "  module foo;\n"
+       "    assign a1 = 1'b0;\n"
+       "  endmodule;\n"
+       "endmodule"},
+
+      // declaration and assignement separated by module block
+      {"module m;\n"
+       "  wire a1;\n"
+       "  module foo;\n"
+       "  endmodule;\n"
+       "  assign a1 = 1'b0;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  wire a1;\n"
+       "  module foo;\n"
+       "    assign a1 = 1'b0;\n"
+       "  endmodule\n"
+       "  assign a1 = 1'b0;\n"
+       "endmodule"},
+
+      // overlapping net
+      {"module m;\n"
+       "  wire a1;\n"
+       "  module foo;\n"
+       "    wire a1;\n"
+       "    assign a1 = 1'b0;\n"
+       "  endmodule\n"
+       "  assign a1 = 1'b0;\n"
+       "endmodule"},
+
+      // multiple declarations
+      {"module m;\n"
+       "  wire a0, a1;\n"
+       "  assign a0 = 1'b0;\n"
+       "  assign a1 = 1'b1;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  wire a0, a2;\n"
+       "  assign a0 = 1'b0;\n"
+       "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+       "endmodule"},
+
+      // multiple net assignments
+      {"module m;\n"
+       "  assign ", {kToken, "a"}, " = b, ", {kToken, "c"}, " = d;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  assign ", {kToken, "a1"}, " = 1'b0;\n"
+       "  wire a1;\n"
+       "endmodule"},
+
+      // out-of-order
+      {"module m;\n"
+       "  assign ", {kToken, "a1"}, " = 1'b0;\n"
+       "  wire a1;\n"
+       "  assign a1 = 1'b1;\n"
+       "endmodule"},
+
+      // concatenated
+      {"module m;\n"
+       "  assign {", {kToken, "a"}, "} = 1'b0;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  assign {", {kToken, "a"}, ",", {kToken, "b"}, "} = 2'b01;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  assign {", {kToken, "a"}, ",", {kToken, "b"}, ",", {kToken, "c"}, "} = 3'b010;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  wire b;\n"
+       "  assign {", {kToken, "a"}, ", b,", {kToken, "c"}, "} = 3'b010;\n"
+       "endmodule"},
+
+      // out-of-scope
+      {"module m;\n"
+       "  module foo;\n"
+       "    wire a1;\n"
+       "  endmodule\n"
+       "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  module foo;\n"
+       "    wire a1;\n"
+       "    assign a1 = 1'b0;\n"
+       "  endmodule\n"
+       "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  wire a1;\n"
+       "  module foo;\n"
+       "    assign a1 = 1'b0;\n"
+       "  endmodule\n"
+       "  assign a1 = 1'b1;\n"
+       "endmodule"},
```

Comment:
This last one is the same as the one I commented earlier. One of the two should be removed.